gateway-service-impl/src/main/java/org/hypertrace/gateway/service/common/datafetcher/QueryServiceEntityFetcher.java

```diff
@@ -354,6 +346,22 @@ public EntityFetcherResponse getEntitiesAndAggregatedMetrics(
     return new EntityFetcherResponse(entityBuilders);
   }
 
+  private void adjustLimitAndOffset(QueryRequest.Builder builder, int limit, int offset) {
+    // If there is more than one groupBy column, we cannot set the same limit that came
+    // in the request since that might return less entities than needed when the same
+    // entity has different values for the other group by columns. Example: A service entity's
+    // name changes and that will now have two different names.
```

Comment:
Based on this example, I think its more a problem in having mutable fields(like api_name, service_name) in pinot
If that is the only case we are trying to solve, may be a better(but requiring effort) approach is to not add the api_name in the group by clause and implement a custom aggregation function like a LATEST(not entirely sure if this is possible too)

Based on https://docs.pinot.apache.org/developers/developers-and-contributors/extending-pinot/custom-aggregation-function, it looks possible, but needs more research

Reply:
We discussed about this with @aaron-steinfeld and @tim-mwangi This was one of the ideas I had too and I've even asked in Pinot community about the feasibility of `LATEST()`. See https://apache-pinot.slack.com/archives/CDRCA57FC/p1602265559126800

I'll log a story to continue the research.